content/blog/unsafe-string-interning-in-go.md

```diff
@@ -0,0 +1,345 @@
+the weak reference back to a `*Value` by using `unsafe.Pointer` conversion, and
+mark the value as resurrected so that any finalizers which run concurrently will
+know that this value cannot be removed from the pool.
+
+```go
+	var v *Value
+	addr, ok := val[cmpVal]; ok {
+		// Value is already interned in the pool. In case a finalizer runs in
+		// the near future, we must "resurrect" this value to bring it back
+		// to the active part of its lifecycle.
+		v = (*Value)((unsafe.Pointer)(addr))
+		v.resurrected = true
+	}
+	if v != nil {
+		return v
+	}
+```
+
+If the value did not already exist in the pool, we can create it now and intern
+it there by adding it to the map of weak references. Note that the
+`v.resurrected` field defaults to `false`, meaning nobody has bothered to
+retrieve this value from the pool as of yet. We also install the finalizer for
+`*Value`.
+
+```go
+	// Value must be created now and then interned. We store a uintptr to
+	// maintain a weak reference so that the value can eventually be garbage
+	// collected if it isn't used for a period of time.
+	v = &Value{cmpVal: cmpVal}
+	val[cmpVal] = uintptr(unsafe.Pointer(v))
+	runtime.SetFinalizer(v, finalize)
+```
+
+Finalizers in Go come with [many
+caveats](https://golang.org/pkg/runtime/#SetFinalizer) but we make use of them
+here in order to ensure that weak references in the map can be deleted when they
+are no longer needed, and thus the underlying `*Value` can be garbage collected
+from the heap at some point in the future.
+
+To avoid a race, we must verify the `resurrected` field and set another
+finalizer if another caller managed to fetch this `*Value` using `Get`
+concurrently. If the value was not resurrected before the finalizer runs, we are
+able to finally free it from the weak reference pool.
+
+```go
+// finalize is a Go finalizer for *Value.
+func finalize(v *Value) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if v.resurrected {
+		// We lost the race. Somebody resurrected it while we were about to
+		// finalize it. Try again next round.
+		v.resurrected = false
+		runtime.SetFinalizer(v, finalize)
+		return
+	}
+	delete(val, v.cmpVal)
+}
+```
+
+Because of the interactions here between weak references, finalizers, and
+possible concurrent calls to `Get`, it may take up to three garbage collection
+cycles to remove a `*Value` from the pool and free it.
+
+1. the finalizer runs and the `resurrected` sentinel is cleared
+1. the finalizer runs again and deletes the value from the weak reference pool
+1. the garbage collector is able to free the associated `*Value` in the heap
+
+And with that, we've successfully implemented unsafe string interning in Go
+using weak references and finalizers! To see the completed implementation and
+for more technical details and background, [check out the code in the
+`go4.org/intern`](https://github.com/go4org/intern/blob/main/intern.go)
+repository.
+
```

Comment:
There's a subtlety that you should mention at some point: It's not exactly guaranteed that two identical values will have the same `*Value`. It's only guaranteed that you won't be able to observe two different values at the same time. :) Imagine that you intern "foo". Then you stop using foo's `*Value`. It gets GC'd. Then later you intern `foo` again. You'll get a new and (probably) different `*Value`--the old one got GC'd. But that's OK, because you can't compare the old vs the new, because you (by definition) have gotten rid of the old one.

Here's the twist. If *your* code plays unsafe games with `*Value`s, thereby hiding them from the GC and thus from package intern, it is theoretically possible that you could break package intern's strong identity guarantee (same pointer means identical value). Unsafe is, well, unsafe.

Reply:
The main way I can think of getting into this bad place would be if you tried to persist *Values to a database somehow. But that's another phrasing of Josh's "if you hide these pointers from the GC, the package invariants break".